File: app/src/main/java/com/savvasdalkitsis/gameframe/draw/view/PaletteViewHolder.java
package com.savvasdalkitsis.gameframe.draw.view;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.savvasdalkitsis.gameframe.R;
import com.savvasdalkitsis.gameframe.draw.model.Palette;

class PaletteViewHolder extends RecyclerView.ViewHolder {

    private final PaletteView paletteView;
    private final TextView title;
    private final View edit;
    private final View delete;
    private final View controls;

    PaletteViewHolder(ViewGroup parent, boolean editable) {
        super(LayoutInflater.from(parent.getContext()).inflate(R.layout.view_palette_view, parent, false));
        title = (TextView) itemView.findViewById(R.id.view_palette_title);
        paletteView = (PaletteView) itemView.findViewById(R.id.view_palette_thumbnail);
        paletteView.setThumbnailMode();
        edit = itemView.findViewById(R.id.view_palette_edit);
        delete = itemView.findViewById(R.id.view_palette_delete);
        controls = itemView.findViewById(R.id.view_palette_controls);
        setEditable(editable);
    }

    private void setEditable(boolean editable) {
        controls.setVisibility(editable ? View.VISIBLE : View.GONE);
    }

    void setDeletable(boolean deletable) {
        delete.setVisibility(deletable ? View.VISIBLE : View.GONE);
    }

    public void bind(Palette palette) {
        title.setText(palette.getTitle());
        paletteView.bind(palette);
    }

    void clearListeners() {
        setOnClickListener(null);
        setOnItemDeletedListener(null);
    }

    void setOnClickListener(View.OnClickListener onClickListener) {
        itemView.setOnClickListener(onClickListener);
    }

    void setOnItemDeletedListener(OnItemDeletedListener onItemDeletedListener) {
        delete.setOnClickListener(v -> onItemDeletedListener.onItemDeleted());
    }

    void setOnPaletteEditClickedListener(OnPaletteEditClickedListener onPaletteEditClickedListener) {
        edit.setOnClickListener(v -> onPaletteEditClickedListener.onPaletteEditClicked());
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/draw/view/PaletteViewHolder.kt
package com.savvasdalkitsis.gameframe.feature.draw.view

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.savvasdalkitsis.gameframe.R
import com.savvasdalkitsis.gameframe.feature.draw.model.Palette
import com.savvasdalkitsis.gameframe.infra.kotlin.Action

internal class PaletteViewHolder(parent: ViewGroup, editable: Boolean) : RecyclerView.ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.view_palette_view, parent, false)) {

    private val paletteView: PaletteView = itemView.findViewById(R.id.view_palette_thumbnail)
    private val title: TextView = itemView.findViewById(R.id.view_palette_title)
    private val edit: View = itemView.findViewById(R.id.view_palette_edit)
    private val delete: View = itemView.findViewById(R.id.view_palette_delete)
    private val controls: View = itemView.findViewById(R.id.view_palette_controls)

    init {
        paletteView.setThumbnailMode()
        setEditable(editable)
    }

    private fun setEditable(editable: Boolean) {
        controls.visibility = if (editable) View.VISIBLE else View.GONE
    }

    fun setDeletable(deletable: Boolean) {
        delete.visibility = if (deletable) View.VISIBLE else View.GONE
    }

    fun bind(palette: Palette) {
        title.text = palette.title
        paletteView.bind(palette)
    }

    fun clearListeners() {
        setOnClickListener(null)
        setOnItemDeletedListener(null)
    }

    fun setOnClickListener(onClickListener: Action?) {
        itemView.setOnClickListener { onClickListener?.invoke() }
    }

    fun setOnItemDeletedListener(onItemDeletedListener: Action?) {
        delete.setOnClickListener { onItemDeletedListener?.invoke() }
    }

    fun setOnPaletteEditClickedListener(onPaletteEditClickedListener: Action) {
        edit.setOnClickListener { onPaletteEditClickedListener() }
    }
}